examples/cis2-multi/src/lib.rs

```diff
@@ -300,6 +398,119 @@ pub struct PermitParamPartial {
     signer:    AccountAddress,
 }
 
+/// The parameter type for the contract function `setPaused`.
+#[derive(Serialize, SchemaType)]
+#[repr(transparent)]
+struct SetPausedParams {
+    /// Specifies if contract is paused.
+    paused: bool,
+}
+
+/// The parameter type for the contract function `upgrade`.
+/// Takes the new module and optionally an entrypoint to call in the new module
+/// after triggering the upgrade. The upgrade is reverted if the entrypoint
+/// fails. This is useful for doing migration in the same transaction triggering
+/// the upgrade.
+#[derive(Serialize, SchemaType)]
+pub struct UpgradeParams {
+    /// The new module reference.
+    pub module:  ModuleReference,
+    /// Optional entrypoint to call in the new module after upgrade.
+    pub migrate: Option<(OwnedEntrypointName, OwnedParameter)>,
+}
+
+/// The parameter for the contract function `grantRole` which grants a role to
+/// an address.
+#[derive(Serialize, SchemaType)]
+pub struct GrantRoleParams {
+    /// The address that has been its role granted.
+    pub address: Address,
+    /// The role that has been granted to the above address.
+    pub role:    Roles,
+}
+
+/// The parameter for the contract function `revokeRole` which revokes a role
+/// from an address.
+#[derive(Serialize, SchemaType)]
+pub struct RevokeRoleParams {
+    /// The address that has been its role revoked.
+    pub address: Address,
+    /// The role that has been revoked from the above address.
+    pub role:    Roles,
+}
+
+/// A struct containing a set of roles granted to an address.
+#[derive(Serial, DeserialWithState, Deletable)]
+#[concordium(state_parameter = "S")]
+struct AddressRoleState<S> {
+    /// Set of roles.
+    roles: StateSet<Roles, S>,
+}
+
+/// Enum of available roles in this contract.
+#[derive(Serialize, PartialEq, Eq, Reject, SchemaType, Clone, Copy, Debug)]
+pub enum Roles {
+    /// Admin role.
+    ADMIN,
+    /// Upgrader role.
+    UPGRADER,
+    /// Blacklister role.
+    BLACKLISTER,
+    /// Pauser role.
+    PAUSER,
```

Comment:
**Suggestion (non-blocking):** Consider adding `MINTER` role as well, but have it commented out.